### ISSUE TITLE
"retrying" enqueue

### ISSUE THREAD

**Issue by MrNghia123** [SGrondin/bottleneck#133]
Thanks for great library.
In my project I need to use highWater together with strategy OVERFLOW. When I put new task to queue, if the queue reject it, it needs to be retried after e.g. 1 second.
There is not much information in the documentation about the behaviour of the submit() or schedule() functions in the case the task is dropped out, so I had to played around with the lib and found the following:
1/ At dropped out, the submit() callback is called with an error of type Bottleneck.BottleneckError saying the task was dropped out. However if the insert was successful, the callback won't be called until the task finish executing, therefore I have nowhere to put my retry logic.
2/ It seems submit() return a Promise, which resolve immediately after the submit attempt with a parameter value TRUE when the task was dropped out, and FALSE when the task was added successfully. I found this behaviour documented nowhere. However after a few test runs it seems to be consistent, so I build my retry logic around it.

Please help to check if the logic is right, and consider adding more information in the documentation. 

```javascript
const Bottleneck = require('bottleneck/es5')
// const logger=require('../log')('test_bottleneck')

const job = function (delay, i, cb) {
  setTimeout(()=>{
    console.log('done', i, ' after ', delay)
    if (i==4)
      cb(new Error('Oops'))
    else 
      cb(null,'Done')
  }, delay)
}

const options = {
  maxConcurrent: 1,
  highWater: 5,
  strategy: Bottleneck.strategy.OVERFLOW
}

limiter = new Bottleneck(options)

async function run () {
  for (let i=0; i<10; i++) {
    await new Promise((resolve, reject) => {
      const wrapper = () => {
        limiter.submit(job, 2000, i, (err, res)=>{
          if (err instanceof Bottleneck.BottleneckError) {
            //
            console.warn( 'Job', i, 'Bottleneck error', err.message)
          } else {
            //Task was run, handle error normally
            if (err) {
              console.error( 'Job', i, 'ended', err)
            } else {
              console.info( 'Job', i, 'ended successfully', res)
            }

          }
        })
        .then((droppedOut)=>{
          if (droppedOut) {
            console.warn('Job' , i, ' dropped out of queue, retrying in 1s')
            setTimeout(wrapper, 1000)
          } else {
            //job enqueued
            resolve()
          }
        })
        .catch(err=>{
          console.log(err)
        })
        console.log('try enqueue job ', i)
      }
      wrapper()
    })
  }
}

run ()
```

**Comment by SGrondin** [comment]
Wow I'm impressed by your reverse engineering!

>1/ At dropped out, the submit() callback is called with an error of type Bottleneck.BottleneckError saying the task was dropped out. However if the insert was successful, the callback won't be called until the task finish executing, therefore I have nowhere to put my retry logic.

This is 100% correct.

2/ It seems submit() return a Promise, which resolve immediately after the submit attempt with a parameter value TRUE when the task was dropped out, and FALSE when the task was added successfully. I found this behaviour documented nowhere. However after a few test runs it seems to be consistent, so I build my retry logic around it.

This is also 100% correct. This "secret" behavior is undocumented for 2 reasons:
- I want to keep `submit()` and `schedule()` as identical as possible. I try to not add features to one that I can't add to the other.
- Documenting this secret behavior means making it official, which means finding a way to explain it. It's not easy and I thought it added too much complexity to the docs with no benefit. I didn't realize your use case existed.

Many of the highwater/strategy tests use this behavior, so it's safe to use.

I've reviewed your code and it looks good to me 👍 . I like the recursive async wrapper.

I think you would be able to use the `dropped` event to accomplish the same thing. This event passes an argument that contains `{ task, options, args }` and it's safe to then directly call `schedule()` like this, even if you used `submit()` to queue the job:

```javascript
limiter.on('dropped', ({ task, args, options}) => {
  limiter.schedule(options, task, ...args)
})
```

The problem is it makes it hard to then let the original caller know that their job is now in a totally different promise.
